code revisions

mojavelinux opened the pull request:
- rename read_paragraph to read_paragraph_lines
- consolidate read_paragraph_lines using procs (aka lambdas)
- change argument order of read_paragraph_lines; use options hash
- rename quote_lines? to blockquote?
- change argument order of blockquote?
- code style

mojavelinux commented:
I took this in a slightly different direction by moving the blocks to constants that we can reuse. That way, we can setup the read_lines_until method once and handle a whole bunch of different permutations. I also verified that we slightly improve the speed of the processor, which is always a goal :) Let me know what you think.

mojavelinux commented on lib/asciidoctor/parser.rb:
```diff
@@ -944,20 +943,18 @@ def self.next_block(reader, parent, attributes = {}, options = {})
+  def self.read_paragraph_lines reader, break_at_list, opts = {}
+    opts[:break_on_blank_lines] = true
+    opts[:break_on_list_continuation] = true
+    opts[:preserve_last_line] = true
```
It's faster here to use explicit assignments rather than a Hash merge.

mojavelinux commented:
How do these changes look to you? Ready to move forward?

Mogztter commented:
I do like the lambda "constants" but at first the following line a bit hard to read (and I don't like the `nil`):
```ruby
&(break_at_list ?
       (Compliance.block_terminates_paragraph ? StartOfBlockOrListProc : StartOfListProc) :
       (Compliance.block_terminates_paragraph ? StartOfBlockProc : nil))
```

Otherwise it's really nice

mojavelinux commented:
I've updated it to (hopefully) be more clear. Specifically, I've added an assignment to a local variable to help define the function of the argument.

I'll admit the nil is somewhat strange, but as it turns out, `&nil` is the equivalent to not passing a block, so it's a convenient way for us to avoid a separate logic branch.

I did a lot of benchmarking and I worked out that this is actually the fastest way to run this code.

The Asciidoctor parser code is always going to look a bit strange because it tries to delegate as much work as possible to the Ruby native code (C or the JVM). The more statements that are written in Ruby, the slower the parser becomes. You can sort of think of it like using a lot of database stored procedures. We are pushing the logic to where it runs the fastest. Unfortunately, that's the trade-off of using Ruby. It's easy to make it slow and hard to make it fast. But it's also a fun challenge. And we continue to chip away at the parse time.

mojavelinux commented:
I could change the `nil` to a constant named `Noop` (or `NoBlock`) that is assigned a value of `nil`. wdyt?

Mogztter commented:
> I could change the nil to a constant named Noop (or NoBlock) that is assigned a value of nil. wdyt?

If it's not slower then yes it could be nice :)

mojavelinux commented:
Updated!

Mogztter commented:
Perfect! :+1:
